utils/logalloc.cc

```diff
@@ -1032,6 +1052,33 @@ void segment_pool::on_memory_allocation(size_t size) {
     _stats.memory_allocated += size;
 }
 
+void segment_pool::increase_latency_reserve_counter(size_t size) {
+    if (_latency_reserve_counter + size > _latency_reserve_threshold + _latency_reserve_antithrashing_margin) [[unlikely]] {
+        assert(size <= _latency_reserve_threshold);
+        segment* seg = new_segment(nullptr);
```

Comment:
This is where I don't understand how it works. new_segment() will have to (eventually) call compact_and_evict(), and that has no run-time guarantee.

Reply:
Due to the reserve, `compact_and_evict()` will evict at most k segments worth of memory before a segment is released from the reserve. And when it's released, `compact_and_evict()` will succeed and stop.

Reply:
Ah, it artificially lowers occupancy in this calculation:

```c++
        if (r->is_evictable() && r->occupancy().used_space() >= max_used_space_ratio_for_compaction * r->occupancy().total_space()) {
            reclaim_from_evictable(*r, target_mem, preempt);
        } else {
            r->compact();
        }
```

but, what places a boundary on the work done by r->compact()?

Reply:
> what places a boundary on the work done by r->compact()?

The 85% compaction threshold makes sure that `r->compact()` makes at least some fixed amount of progress. Since we compact only segments with occupancy <85%, and the max object size is (10%*segment_size+epsilon), then a compaction reduces the amount of unusable holes by at least (5%*segment_size-epsilon). So at most 21 compactions is needed to free a segment. If only small objects are around, then at most 7 compactions, since every compaction frees (15%*segment_size - epsilon).

Compaction is expensive. I wonder if what we would gain and lose by keeping an intrusive list of holes and allocating inside them if possible, instead of allocating only in the active segment and relying on compactions to get rid of holes. On one hand, this should significantly decrease the amount of compactions. On the other hand we would lose some locality (allocating only in the active segment has the nice property that objects allocated at the same time should generally stay close to each other in memory).